File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.py
class Solution:
    def maxSumAfterPartitioning(self, arr: List[int], k: int) -> int:
        
        length = len(arr)
        
        array = [0] * (length + 1)
        
        for i in range(1, length+1):
            maxi = -math.inf
            for j in range(1, min(i, k) + 1):
                maxi = max(maxi, arr[i - j])
                array[i] = max(array[i], array[i - j] + maxi * j)
        
        return array[-1]

File: 1043-partition-array-for-maximum-sum/1043-partition-array-for-maximum-sum.java
class Solution {
    public int maxSumAfterPartitioning(int[] arr, int k) {
        
        int n = arr.length;
        int[] result = new int[n + 1];
        
        for (int i = 1; i <= n; i++) {
            int max = Integer.MIN_VALUE;
            
            for (int j = 1; j <= Math.min(i, k); j++) {
                max = Math.max(max, arr[i - j]);
                result[i] = Math.max(result[i], result[i - j] + max * j);
            }
        }
        
        return result[n];
    }
}
